KAFKA-18844: Stale features information in QuorumController#registerBroker (#18997)

FrankYang0529 opened the pull request:
In https://github.com/apache/kafka/pull/16848, we added `kraft.version` to finalized features and got finalized features outside controller event handling thread. This may make finalized features stale when processing `registerBroker` event. Also, some cases like `QuorumControllerTest.testBalancePartitionLeaders` become flaky cause of outdated MV. This PR moves finalized features back to controller event handling thread to avoid the error.

Reviewers: Ismael Juma <ijuma@apache.org>, Jun Rao <junrao@gmail.com>, Colin P. McCabe <cmccabe@apache.org>, Chia-Ping Tsai <chia7712@gmail.com>

(cherry picked from commit 88a23dab3ea6f76cc32066066149ce7843bf24ab)

junrao reviewed (commented):
@FrankYang0529 : Thanks for the cherry-pick. Are the test failures, especially the following two related? 


```
Build / JDK 8 and Scala 2.12 / org.apache.kafka.controller.QuorumControllerTest.testFenceMultipleBrokers() | 41 sec | 1
-- | -- | --
Build / JDK 21 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testFenceMultipleBrokers()

[Build / JDK 8 and Scala 2.12 / org.apache.kafka.controller.QuorumControllerTest.testFenceMultipleBrokers()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-19058/1/testReport/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_8_and_Scala_2_12___testFenceMultipleBrokers__/)	41 sec	[1](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-19058/1/)
 [Build / JDK 21 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testFenceMultipleBrokers()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-19058/1/testReport/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_21_and_Scala_2_13___testFenceMultipleBrokers__/)
```

FrankYang0529 commented:
Hi @junrao, these test cases can pass with zelu-1.8 and Amazon Corretto 21 on my laptop, so I will trigger CI to check the result again. Thanks.

FrankYang0529 committed:
KAFKA-18844: Stale features information in QuorumController#registerBroker (#18997)

In https://github.com/apache/kafka/pull/16848, we added `kraft.version`
to finalized features and got finalized features outside controller
event handling thread. This may make finalized features stale when
processing `registerBroker` event. Also, some cases like
`QuorumControllerTest.testBalancePartitionLeaders` become flaky cause of
outdated MV. This PR moves finalized features back to controller event
handling thread to avoid the error.

Reviewers: Ismael Juma <ijuma@apache.org>, Jun Rao <junrao@gmail.com>,
Colin P. McCabe <cmccabe@apache.org>, Chia-Ping Tsai
<chia7712@gmail.com>

(cherry picked from commit 88a23dab3ea6f76cc32066066149ce7843bf24ab)

dajac commented:
@FrankYang0529 Do we also need a PR for 4.0 or was it already cherry-picked?

FrankYang0529 commented:
@dajac, I think we can just cherry-picked it, because there is no conflict.

dajac commented:
> @dajac, I think we can just cherry-picked it, because there is no conflict.

Sounds good. I did it.

junrao commented:
Thanks, David and Frank. Closing the PR.

FrankYang0529 commented:
@junrao, there is conflict in 3.9, so I created this PR. In 3.9 source code, the finalized features are still outside controller event handling thread, so I reopen the PR. Thanks.

chia7712 commented:
JDK 17 pass
